### PR TITLE
make ComplexColumn an interface and ExtensionPoint

### PR DESCRIPTION
**Pull request by himanshug**
Motivation:
Druid gives full control over serialization of a complex column to the extension writer. As a result custom extension can have complex columns which could have no GenericIndex or many of them.

Changes:
So, this patch removes dependency of ComplexColumn from GenericIndexed . Also, annotates it with @ExtensionPoint which was just missed originally.

Operational Impact:
None

Test Plan:
No behavioral change is introduced in this patch, it is only slight rearrangement of code. Existing tests cover all of it.

PS: I thought of first writing a proposal but changes are so small I merged the proposal into this PR itself.

### CODE REVIEW

**Comment by himanshug** [comment]
@gianm @jon-wei @jihoonson @clintropolis  can someone review this ?

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java [hunk 1]
```diff
@@ -19,45 +19,29 @@
...
 import org.apache.druid.segment.data.ReadableOffset;
 
 import javax.annotation.Nullable;
 
 /**
-*/
-public class ComplexColumn implements BaseColumn
+ */
+@ExtensionPoint
+public interface ComplexColumn extends BaseColumn
```
Comment (jihoonson) [inline processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: Would you please add some javadoc?

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java [hunk 2]
```diff
@@ -19,45 +19,29 @@
...
-  public int getLength()
-  {
-    return index.size();
-  }
+  @Override
+  void close();
 
   @Override
-  public ColumnValueSelector<?> makeColumnValueSelector(ReadableOffset offset)
+  default ColumnValueSelector makeColumnValueSelector(ReadableOffset offset)
```
Comment (jihoonson) [inline processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: Why did you remove the wildcard?

File: processing/src/main/java/org/apache/druid/segment/column/GenericIndexedBasedComplexColumn.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.column;
+
+import org.apache.druid.segment.data.GenericIndexed;
+
+/**
+*/
```
Comment (jihoonson) [inline processing/src/main/java/org/apache/druid/segment/column/GenericIndexedBasedComplexColumn.java]: Please add a javadoc what this class represents.

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java (hunk 1 above)
Comment (himanshug) [reply processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: added

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java (hunk 2 above)
Comment (himanshug) [reply processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: that was a mistake , thanks.

File: processing/src/main/java/org/apache/druid/segment/column/GenericIndexedBasedComplexColumn.java (hunk 3 above)
Comment (himanshug) [reply processing/src/main/java/org/apache/druid/segment/column/GenericIndexedBasedComplexColumn.java]: added

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java [hunk 4]
```diff
@@ -19,45 +19,34 @@
...
 import org.apache.druid.segment.ObjectColumnSelector;
-import org.apache.druid.segment.data.GenericIndexed;
 import org.apache.druid.segment.data.ReadableOffset;
 
 import javax.annotation.Nullable;
 
 /**
-*/
-public class ComplexColumn implements BaseColumn
+ * This interace represents a complex column and can be implemented by druid extension writer of a custom column
```
Comment (jihoonson) [inline processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: interace -> interface?

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java [hunk 5]
```diff
@@ -19,45 +19,34 @@
...
+ * with arbitrary serialization instead of a custom column that serializes rows of objects serialized using
+ * {@link org.apache.druid.segment.data.GenericIndexed} class which is default implementation of "writeToXXX" methods in
+ * {@link org.apache.druid.segment.serde.ComplexColumnSerializer}. In that case {@link GenericIndexedBasedComplexColumn}
+ * should be used.
+ */
+@ExtensionPoint
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: Do additional column interfaces need to be marked as @ExtensionPoint for this to make sense?

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java (hunk 1 above)
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: Oh sorry, I should mention this in the first place, but would you please add simple descriptions for methods as well? Since this is an extension point now, it would make it clear what their implementation requires.

Comment (himanshug) [reply processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: added more to the best of my knowledge :)

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java (hunk 5 above)
Comment (himanshug) [reply processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: that is good point. I did another pass at things and most of the stuff looks to be marked correctly already. I had to mark ColumnValueSelector hierarchy as @ExtensionPoint .

File: processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java (hunk 4 above)
Comment (himanshug) [reply processing/src/main/java/org/apache/druid/segment/column/ComplexColumn.java]: that was intentional to see if you were actually reading the doc update.

**Comment by clintropolis** [comment]
failing CI though 

```
[ERROR] /home/travis/build/apache/incubator-druid/processing/src/main/java/org/apache/druid/segment/QueryableIndexIndexableAdapter.java:[367,28] unreported exception java.io.IOException; must be caught or declared to be thrown
  exception thrown from implicit call to close() on resource variable 'complexColumn'
```

**Comment by himanshug** [comment]
thanks guys, will fix the build and merge then.
